### ISSUE TITLE
I have a BD, working with which mkvtoolnix eats all memory.

### ISSUE THREAD

**Issue by k0ste** [mbunkus/mkvtoolnix#2020]
Hello.
This is it [A River Runs Through It](http://hdclub.org/details.php?id=28130). Have read comments, I found out I'm not alone.

ArchLinux, mkvmerge v12.0.0 ('Trust / Lust') 64bit, cmd line:
```
/usr/bin/mkvmerge --ui-language en_US --output /mnt/nas/Movies/A_River_Runs_Through_It.mkv --audio-tracks 1 --subtitle-tracks 7 --language 0:und --language 1:rus --language 7:rus '(' /mnt/nas/DOWNLOAD/A.River.Runs.Through.It.1992.BD.REMUX/BDMV/STREAM/00000.m2ts ')' --track-order 0:0,0:1,0:7
```

If you prefer, I can upload this release to ftp (~41 Gbytes).
Thanks.

**Comment by mbunkus** [comment]
I will definitely need access to the disc in question. Therefore I'd like to take you up on your offer. Please upload it to my [FTP server](https://github.com/mbunkus/mkvtoolnix/wiki/FTP-server). Thanks.

**Comment by mkver** [comment]
You might do one thing before you upload a whole BD: Mux only a small part (for which you have enough memory) and then check whether the resulting file has keyframes. You can e.g. use mkvinfo for that*: Let it show you everything and then look at the cues at the end of the file. You should have about one cue entry (corresponding to a keyframe) per second. If you don't then your stream likely doesn't contain so called "recovery point SEI" messages that tell mkvmerge that your I-frames are keyframes. In such a case you can use the ``--engage all_i_slices_are_key_frames`` switch and hope that all I-frames are indeed valid keyframes.
*: Alternatively, mkvextract can extract the cues of a track.

**Comment by k0ste** [comment]
Upload complete.

**Comment by mbunkus** [comment]
This h.264 stream seems to contain only one real key frame, right at the start of the file. Extracting the raw h.264 from the MPEG TS and then importing that into an MP4 file with MP4Box shows just one key frame, no RAP atoms for OpenGOP situations either.

This is something that mkvmerge simply cannot deal with, and judging by how much other stuff I have to do it's highly unlikely I'll ever invest time here. This means that the issue with this disc will remain unfixed. Sorry.

**Comment by k0ste** [comment]
it is possible to handle files like that? For example,  error: "I can't work with %s (see github #2020)", instead of eating all memory.

**Comment by mbunkus** [comment]
Not easily, no.

**Comment by remuxer32** [comment]
@k0ste 

If you still want to put your movie to `.mkv` (but without eating all memory, of course), you should use MakeMKV. MakeMKV fully supports Blu-ray Discs, as well as DVDs. It is able to show you all audio tracks and subtitles which are present on your disc. You can easily select your wanted tracks to keep.

About your movie: that is not a full Blu-ray Disc; that's a BD-Remux. A remux is a lossless copy (video bitrate is intact) but without the main menu for example, and without important folders (files) which are always present on the original discs. To put it simply: it's only a main movie, nothing more. Blu-ray remuxes are also met in `.mkv`. Your remux is in `.m2ts`, of course. I personally prefer `.mkv` as it's more flexible.

But to the point: MakeMKV is able to handle BD-Remuxes as well. All you have to do is to add the `index.bdmv` (or `MovieObject.bdmv`) file from the `BDMV` folder — MakeMKV will start opening your disc. 

Useful links:

- http://www.makemkv.com/download/

Here's a guide how to install MakeMKV on Linux:

- http://www.makemkv.com/forum2/viewtopic.php?f=3&t=224

For Arch Linux directly:

- https://aur.archlinux.org/packages/makemkv/

Good luck.

**Comment by mkver** [comment]
@remuxer32: Does MakeMKV really analyze the bitstream to find out which of the I-frames is a valid keyframe? Or does it just declare every I-frame to be a keyframe (just as mkvmerge does with ``--engage all_i_slices_are_key_frames`` command line switch -- as already mentioned, this is something k0ste can try)? Or do you mean that it muxes to mkv without excessive memory consumption, but without any cuepoints (except for the very first frame), making the file quite unseekable?

**Comment by remuxer32** [comment]
I meant only muxing to `.mkv` without huge memory consumption; that's it, nothing else. As I had experience with several BD-Remuxes in the past, I decided to share my experience (meaning: MakeMKV recognizes/handles such lossless copies as well, not only full Blu-ray Discs). Of course I haven't tested that particular BD-Remux, hence I cannot be sure about the final result, but I just gave a hint what he can do.

Obviously I'm a fan of muxing directly from a full Blu-ray Disc, not from a copy. Most remuxers (people who make such lossless copies) use tsMuxeR (which is not supported anymore, and which has tons of bugs undoubtedly), hence any side effects can occur, but that's not my problem. I recommed not to use outdated software (vide tsMuxeR) for making Blu-ray copies in `.m2ts`, but always use such tools like MKVToolNix or MakeMKV (which are still being developed) in order to put a movie to the Matroska file format (more flexible than MPEG TS).
